manticore/core/smtlib/solver.py

```diff
@@ -418,7 +418,7 @@ def can_be_true(self, constraints: ConstraintSet, expression=True):
                 return expression
             else:
                 # if True check if constraints are feasible
-                self._reset(constraints)
+                self._reset(constraints.to_string())
```

Comment:
Previously, we could pass `ConstraintSet`s around without needing to explicitly render them to strings as the f-string in `_send` would implicitly call `__str__`. Just offhand I can't see why one way of doing this would be inherently better than the other, but I'm guessing that this way is more convenient from a type standpoint. Can you elaborate?

Reply:
Certainly!  [Elsewhere in this pull request](https://github.com/trailofbits/manticore/pull/1668/files#r408833413), the type of the `Z3Solver._reset()` method was annotated explicitly as _only_ accepting an optional string argument of constraints, and `Z3Solver._send()` had its type annotated as taking a string (rather than _either a string or ConstraintSet_).

My thinking is that explicit `Union` types are more difficult to reason about that single types or interface types — more cases to worry about in the implementation!

In light of the `_send()` and `_reset()` changes, the call sites of those methods need to be made consistent.  Hence the addition of the explicit `.to_string()` calls.  `mypy` will let you know if you've crossed the streams.

Reply:
> My thinking is that explicit Union types are more difficult to reason about that single types or interface types

Okay, I agree. Since `_send` and `_reset` aren't user-facing this shouldn't break anything either.